R/slides.R

```diff
@@ -0,0 +1,331 @@
+            # and reorder the slideCatalog.
+            new_order <- moveLastElement(seq_len(n_slides + 1), i)
+            reorderSlides(x, new_order)
+            with_consent(delete(x[[length(x)]]))
+        }
+        invisible(refresh(x))
+    }
+)
+
+#' Move and delete last element of a vector
+#' This moves the last element of a vector to an index, then deletes the last
+#' element, it is broken out for testing purposes
+#'
+#' @param v a vector
+#' @param idx The index to move the last element to.
+#'
+#' @return a vector
+moveLastElement <- function(v, idx){
+    v[idx] <- v[length(v)]
+    out <- v[1:(length(v) -1)]
+    return(out)
+}
+
+#' Reorder slides in a crunchdeck
+#'
+#' @param x A SlideCatalog
+#' @param order The numeric order for slides to be reordered to.
+#'
+#' @return A SlideCatalog
+reorderSlides <- function(x, order) {
+    url <- paste0(self(x), "flat")
+    payload <- crGET(url)
+    payload$graph <- payload$graph[order]
+    crPATCH(url, body = toJSON(payload))
+    return(refresh(x))
+}
+
+setMethod("delete", "CrunchSlide", function (x, ...) {
+    if (!askForPermission(paste0("Really delete slide ", dQuote(title(x)), "?"))) {
+        halt("Must confirm deleting CrunchSlide")
+    }
+    out <- crDELETE(self(x), drop = dropCache(absoluteURL("../", self(x))))
+    invisible(out)
+})
+
+# CrunchSlide -------------------------------------------------------------------
+
+setMethod("show", "CrunchSlide", function(object){
+    out <- cubes(object)
+    names(out) <- title(object)
+    print(out)
+})
+
+# TODO: Find out what the mandatory display settings should be for the app then
+# change this list to reflect those settings.
+DEFAULT_DISPLAY_SETTINGS <- list(
+    percentageDirection = "colPct",
+    showEmpty = FALSE,
+    showMean = FALSE,
+    vizType = "table",
+    countsOrPercents = "percent",
+    decimalPlaces = 1L,
+    showSignif = TRUE,
+    currentTab = 0L
+)
+
+#' Append a new slide to a Crunch Deck
+#'
+#' @param deck A Crunch Deck
+#' @param query A formula definition of a query to be used by the slide. This is
+#' similar to CrunchCube query
+#' @param display_settings (optional) A list of display settings, if om
+#' @param title The slide's title
+#' @param subtitle The slide's subtitle
+#' @param ... Further options to be passed on to the API
+#'
+#' @return
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' newSlide(
+#'     main_deck,
+#'     ~cyl + wt,
+#'     title = "Cyl and Weight",
+#'     subtitle = 2017 Data"
+#'     )
+#' }
+newSlide <- function(deck,
+                     query,
+                     display_settings = list(),
+                     title = "",
+                     subtitle = "",
+                     ...) {
+    ds <- loadDataset(datasetReference(deck))
+    if (inherits(query, "formula")) {
+        query <- list(query)
+    }
+
+    settings <- modifyList(display_settings, DEFAULT_DISPLAY_SETTINGS)
```

Comment:
Are the arguments to `modifyList` flipped here? Wouldn't this make the defaults always be used no matter if they were specified above?

```
> modifyList(list(a=1), list(a=4, b=2))
$a
[1] 4

$b
[1] 2
```